File: 20. Valid Parentheses/Valid Parentheses.py
"""
My Solution
"""
class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        stack = []
        parentheses = {
            '(': ')',
            '[': ']',
            '{': '}'
        }

        for c in s:
            if c in parentheses:
                stack.append(parentheses[c])
            else:
                pop_value = stack.pop() if stack else None
                if pop_value != c:
                    return False
        return True if not stack else False

File: 20. Valid Parentheses/Valid Parentheses.js
/**
 * @param {string} s
 * @return {boolean}
 */

const map = {
    "(" : ")",
    "{" : "}",
    "[" : "]",
}

var isValid = function(s) {
    const stk = [];
    
    for (const k of s) {
        if (map[k]) {
            stk.push(map[k])
        } else {
            if (stk.pop() !== k) {
                return false;
            }
        }
    }

    return stk.length === 0;
};